sql/plan/json_table.go

```diff
@@ -0,0 +1,226 @@
+	return nil
+}
+
+func (j *jsonTablePartitionIter) Next(ctx *sql.Context) (sql.Partition, error) {
+	if j.pos >= len(j.keys) {
+		return nil, io.EOF
+	}
+
+	key := j.keys[j.pos]
+	j.pos++
+	return &jsonTablePartition{key}, nil
+}
+
+type jsonTableRowIter struct {
+	rows []sql.Row
+	pos  int
+}
+
+var _ sql.RowIter = &jsonTableRowIter{}
+
+func (j *jsonTableRowIter) Next(ctx *sql.Context) (sql.Row, error) {
+	if j.pos >= len(j.rows) {
+		return nil, io.EOF
+	}
+	row := j.rows[j.pos]
+	j.pos++
+	return row, nil
+}
+
+func (j *jsonTableRowIter) Close(ctx *sql.Context) error {
+	return nil
+}
+
+type JSONTable struct {
+	name     string
+	dataExpr sql.Expression
+	path     string
+	schema   sql.PrimaryKeySchema
+	colPaths []string
+}
+
+var _ sql.Table = &JSONTable{}
+var _ sql.Node = &JSONTable{}
+var _ sql.Expressioner = &JSONTable{}
+
+// Name implements the sql.Table interface
+func (t *JSONTable) Name() string {
+	return t.name
+}
+
+// String implements the sql.Table interface
+func (t *JSONTable) String() string {
+	return t.name
+}
+
+// Schema implements the sql.Table interface
+func (t *JSONTable) Schema() sql.Schema {
+	return t.schema.Schema
+}
+
+// Partitions implements the sql.Table interface
+func (t *JSONTable) Partitions(ctx *sql.Context) (sql.PartitionIter, error) {
+	// TODO: this does nothing
+	return &jsonTablePartitionIter{
+		keys: [][]byte{{0}},
+	}, nil
+}
+
+// PartitionRows implements the sql.Table interface
+func (t *JSONTable) PartitionRows(ctx *sql.Context, partition sql.Partition) (sql.RowIter, error) {
+	return t.RowIter(ctx, nil)
+}
+
+// Resolved implements the sql.Resolvable interface
+func (t *JSONTable) Resolved() bool {
+	return t.dataExpr.Resolved()
+}
+
+// Children implements the sql.Node interface
+func (t *JSONTable) Children() []sql.Node {
+	return nil
+}
+
+// RowIter implements the sql.Node interface
+func (t *JSONTable) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	// TODO: need to resolve function calls like concat()
+	// data must evaluate to JSON string
+	data, err := t.dataExpr.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	strData, ok := data.(string)
+	if !ok {
+		return nil, fmt.Errorf("invalid data type for JSON data in argument 1 to function json_table; a JSON string or JSON type is required")
+	}
+
+	// Parse data as JSON
+	var jsonData interface{}
+	if err := json.Unmarshal([]byte(strData), &jsonData); err != nil {
+		return nil, err
+	}
+
+	// Get data specified from initial path
+	var jsonPathData []interface{}
+	if rootJSONData, err := jsonpath.JsonPathLookup(jsonData, t.path); err == nil {
+		if data, ok := rootJSONData.([]interface{}); ok {
+			jsonPathData = data
+		} else {
+			jsonPathData = []interface{}{rootJSONData}
+		}
+	} else {
+		return nil, err
+	}
+
+	// Create new RowIter
+	itr := &jsonTableRowIter{
+		rows: []sql.Row{},
+	}
+
+	// Fill in table with data
+	for colIdx, p := range t.colPaths {
```

Comment:
Don't do any of this iteration in RowIter, and don't store it all in memory

Do it one row at a time in iter.Next()